docs/rewriter/rewrite_patterns.md

```diff
@@ -0,0 +1,113 @@
+3. `pattern_rewrite_rules` : `(Optional)` This paramter is used to pass rewrite rules based on a provided replacement pattern. For the purpose of this tutorial, we will be using only this parameter in conjuction with `model`. This parameter is of type `Sequence[PatternRewriteRule]`.
+
+Note: `pattern_rewrite_rules` takes a sequence of `PatternRewriteRule` types, so if only a singular rewrite rule is to be passed, it needs to passed as part of a sequence.
+
+The snippet below below demonstrates how to use the `rewriter.rewrite` call for the rewrite rule created above:
+
+```{literalinclude} examples/erfgelu.py
+:lines: 88-91
+```
+
+The graph (on the left) consists of the target pattern before the rewrite rule is applied. Once the rewrite rule is applied, the graph (on the right) shows that the target pattern has been succesfully replaced by a GELU node as intended.
+
+![target_pattern](examples/img/erfgelu_01.png) ![replacement_pattern](examples/img/erfgelu_02.png)
+
+
+## Utilizing match_condition parameter for pattern-matching
+
+This section talks about how to utilize the `match_condition` parameter. The `match_condition` parameter checks if the pattern matches the target pattern with certain constraints in consideration.
+
+Let us consider a model which consists of the following pattern.
+
+![target_pattern](examples/img/broadcast_01.png)
+
+Based on the [ONNX Matmul spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#MatMul), onnx Matmul behaves like numpy.matmul and also follows numpy broadcasting. So in this particular pattern if matmul broadcasting is enough, then we don't need the reshapes. To validate this, we need to check the following:
+
+1. Input shapes check: input_a and input_b should be broadcastable
+2. Output shape check: shape_c should be the same as the output shape from the matmul(input_a, input_b)
+
+If the above are true, then we don't need the reshapes and we can eliminate them using a pattern based rewrite. In order to validate whether matmul broadcast is sufficient, we write a condition checking function as follows:
+
+```{literalinclude} examples/broadcast_matmul.py
+:pyobject: check_if_need_reshape
+```
+
+Once we have the match_condition function, we can write a target pattern and replacement pattern in a similar way to the first example.
+
+```{literalinclude} examples/broadcast_matmul.py
+:pyobject: two_reshapes_matmul_reshape_pattern
+```
+
+```{literalinclude} examples/broadcast_matmul.py
+:pyobject: matmul_with_two_shape_inputs
+```
+
+With all the necessary components in place, the pattern rewrite rule is created and then the `rewriter.rewrite` is called to apply the rewrite.
+
+```{literalinclude} examples/broadcast_matmul.py
+:lines: 199-203
+```
+```{literalinclude} examples/broadcast_matmul.py
+:lines: 204-207
```

Comment:
Is it robust to specify line in here? I suppose we don't get notified if the source file is changed.

Reply:
That's a good point, updated to reflect this.